crypto/bn/rsaz_exp_x2.c

```diff
@@ -0,0 +1,438 @@
+        EXPx2 = RSAZ_exp52x20_x2;
+        break;
+    default:
+        goto err;
+    }
+
+    storage = (BN_ULONG *)OPENSSL_malloc(storage_len_bytes + 64);
+    if (NULL == storage)
+        goto err;
+    storage_aligned = (BN_ULONG *)ALIGN_OF(storage, 64);
+
+    /* Memory layout for red(undant) representations */
+    base1_red = storage_aligned;
+    base2_red = storage_aligned + 1 * exp_digits;
+    m1_red    = storage_aligned + 2 * exp_digits;
+    m2_red    = storage_aligned + 3 * exp_digits;
+    rr1_red   = storage_aligned + 4 * exp_digits;
+    rr2_red   = storage_aligned + 5 * exp_digits;
+    res1_red  = storage_aligned + 6 * exp_digits;
+    res2_red  = storage_aligned + 7 * exp_digits;
+
+    /* Convert base_i, m_i, rr_i, from regular to 52-bit radix */
+    to_words52(base1_red,  exp_digits, base1, factor_size);
+    to_words52(base2_red,  exp_digits, base2, factor_size);
+    to_words52(m1_red,  exp_digits, m1,    factor_size);
+    to_words52(m2_red,  exp_digits, m2,    factor_size);
+    to_words52(rr1_red, exp_digits, rr1,   factor_size);
+    to_words52(rr2_red, exp_digits, rr2,   factor_size);
+
+    /* Compute target domain Montgomery converter RR' */
+    memset(res2_red, 0, exp_digits * sizeof(BN_ULONG));
+    set_bit(res2_red, 4 * (DIGIT_SIZE * exp_digits - factor_size));
+    to_words52(res1_red, exp_digits, res2_red, factor_size);
+
+    /* AMM = Almost Montgomery Multiplication */
+    AMM(rr1_red, rr1_red, rr1_red, m1_red, k0_1); /* RR*R */
+    /* RR' = RR*2^k */
+    AMM(rr1_red, rr1_red, res1_red /* holds 2^k value */, m1_red, k0_1);
+    AMM(rr2_red, rr2_red, rr2_red, m2_red, k0_2);
+    AMM(rr2_red, rr2_red, res1_red /* holds 2^k value */, m2_red, k0_2);
+
+    const BN_ULONG *exp[2] = {0};
+    exp[0] = exp1;
+    exp[1] = exp2;
+
+    BN_ULONG k0[2] = {0};
+    k0[0] = k0_1;
+    k0[1] = k0_2;
+
+    /* Call dual (2-exps in parallel) exponentiation kernel */
+    EXPx2(rr1_red, base1_red, exp, m1_red, rr1_red, k0);
+
+    /* Convert rr_i back to regular radix */
+    from_words52(res1, factor_size, rr1_red);
+    from_words52(res2, factor_size, rr2_red);
+
+    ret = 1;
+err:
+    if (NULL != storage) {
+        OPENSSL_cleanse(storage, storage_len_bytes);
+        OPENSSL_free(storage);
+    }
+    return ret;
+}
+
+void extract_multiplier_2x20_win5(BN_ULONG *red_Y,
+                                  const BN_ULONG *red_table,
+                                  int red_table_idx, int tbl_idx);
+/* Uses AVX512VL 256-bit IFMA */
+void RSAZ_ams52x20_x2_256(BN_ULONG *out, const BN_ULONG *a,
+                          const BN_ULONG *m, const BN_ULONG k0[2]);
+void RSAZ_amm52x20_x2_256(BN_ULONG *out, const BN_ULONG *a,
+                          const BN_ULONG *b, const BN_ULONG *m,
+                          const BN_ULONG k0[2]);
+
+void RSAZ_ams52x20_x2_256(BN_ULONG *out, const BN_ULONG *a,
+                          const BN_ULONG *m, const BN_ULONG k0[2])
+{
+    RSAZ_amm52x20_x2_256(out, a, a, m, k0);
```

Comment:
What is the difference between AMM and AMS? I assume AMM stands for "Almost Montgomery Multiplication"

Reply:
AMS stands for Almost Montgomery Squaring. I've added more details into the comments section.